storage/src/robocorp/storage/_client.py

```diff
@@ -0,0 +1,170 @@
+    """No asset with given name/id found."""
+
+
+class AssetUploadFailed(RuntimeError):
+    """There was an unexpected error while uploading an asset."""
+
+
+class AssetsClient:
+    def __init__(self, workspace: str, endpoint: str, token: str):
+        self._endpoint = endpoint
+        self._workspace = workspace
+        self._token = token
+
+        route_prefix = url_join(self._endpoint, "workspaces", self._workspace, "assets")
+        default_headers = {
+            "Authorization": f"Bearer {self._token}",
+            "Content-Type": "application/json",
+        }
+
+        self._client = Requests(route_prefix, default_headers=default_headers)
+
+    def _handle_asset_not_found(self, asset_id: str):
+        """Create an error handler that raises `AssetNotFound` for 404 status codes."""
+
+        def _handle_error(response: Response):
+            try:
+                self._client.handle_error(response)
+            except HTTPError as exc:
+                if exc.status_code == 404:
+                    # NOTE(cmin764): Any `404 Not Found` error won't be retried
+                    #  and will bubble-up from the first request.
+                    raise AssetNotFound(
+                        f"Asset with id {asset_id!r} not found",
+                        status_code=exc.status_code,
+                        reason=exc.reason,
+                    ) from exc
+                raise
+
+        return _handle_error
+
+    def list_assets(self) -> list[Asset]:
+        """Retrieve list of assets."""
+        return self._client.get("").json()
+
+    def create_asset(self, name: str) -> AssetDetails:
+        """Create a new asset."""
+        payload = {"name": name}
+        response = self._client.post("", json=payload)
+        return cast(AssetDetails, response.json())
+
+    def get_asset(self, asset_id: str) -> AssetDetails:
+        """Returns the details of a single asset, including its payload."""
+        path = quote_plus(asset_id)
+        handler = self._handle_asset_not_found(asset_id)
+        response = self._client.get(path, _handle_error=handler)
+        return cast(AssetDetails, response.json())
+
+    def delete_asset(self, asset_id: str):
+        """Delete the given asset."""
+        path = quote_plus(asset_id)
+        handler = self._handle_asset_not_found(asset_id)
+        self._client.delete(path, _handle_error=handler)
+
+    def upload_asset(
+        self,
+        asset_id: str,
+        content: bytes,
+        content_type: Optional[ContentType] = None,
+        wait=True,
+        poll_interval=0.5,
+    ):
+        """Upload an asset payload, and optionally wait for it to finish."""
+        if _estimate_base64_size(len(content)) < DATA_LIMIT:
+            LOGGER.info("Content size under data limit, uploading directly")
+            data_content = base64.b64encode(content).decode("ascii")
+            assert len(data_content) < DATA_LIMIT, "Unexpected data size"
```

Comment:
Instead of asserting, we can just switch to the "full" extensive data upload once we confirm that our assumption was wrong. And the `LOGGER.info("Content size under data limit, uploading directly")` can be placed after this check, so we'd print it once we're sure we're able to upload the data directly, otherwise we fallback to the lengthy upload scenario.

Reply:
The calculation _should_ never be wrong, although I guess it can be juggled around to make it behave more nicely if the code is misbehaving in some way

Reply:
Yes, checking the size, and if is small then actually encoding it for the quick transfer check, then checking it again to make sure is really what we expected. And only then decide if we still go on the quick or long way, without raising any error where we're in control.
Would add more peace to us as it makes the code more resilient to such corner case that should never happen.

Reply:
Actually, since the limit is actually 6MB, the difference between the estimate and reality is at most a few bytes, I'll just remove the assert. The estimation should be plenty to ensure we're within the "safe limits".